## Library import

In [2]:
import pandas as pd
import glob
import os
import json

Pandas: library used to work with tabular data, useful to analyze, manipulate and clean data<br>
Glob: allows to search files in a folder using patterns (e.g. *.csv)<br>
Os: allows you to interact with the operating system (paths, folders, files)<br>
Json: used to read and write data in JSON format

---

## Exercise 1
### Create a single dataframe with the concatenation of all input csv files, adding a column called country


In [3]:
path = r"/Users/nicolo/Desktop/Focs_project/trendingYT"
files = glob.glob(os.path.join(path, "*videos.csv.zst"))

full_df = pd.concat((pd.read_csv(f,compression="zstd",encoding="utf-8",encoding_errors = "replace")
                       .assign(country=os.path.basename(f)[:2]) for f in files),
                    ignore_index=True)

full_df

video_id trending_date  \
0       2kyS6SvSYSE      17.14.11   
1       1ZAPwfrtAFY      17.14.11   
2       5qpjK5DgCt4      17.14.11   
3       puqaWrEC7tY      17.14.11   
4       d380meD0W0M      17.14.11   
...             ...           ...   
375937  l884wKofd54      18.14.06   
375938  IP8k2xkhOdI      18.14.06   
375939  Il-an3K9pjg      18.14.06   
375940  -DRsfNObKIQ      18.14.06   
375941  4YFo4bdMO8Q      18.14.06   

                                                    title  \
0                      WE WANT TO TALK ABOUT OUR MARRIAGE   
1       The Trump Presidency: Last Week Tonight with J...   
2       Racist Superman | Rudy Mancuso, King Bach & Le...   
3                        Nickelback Lyrics: Real or Fake?   
4                                I Dare You: GOING BALD!?   
...                                                   ...   
375937  Enrique Iglesias - MOVE TO MIAMI (Official Vid...   
375938  Jacob Sartorius - Up With It (Official Music V...   
375939                 Anne-Marie - 2002 [Official Video]   
375940  Eleni Foureira - Fuego - Cyprus - LIVE - First...   
375941  KYLE - Ikuyo feat.  2 Chainz & Sophia Black [A...   

                  channel_title  category_id              publish_time  \
0                  CaseyNeistat           22  2017-11-13T17:13:01.000Z   
1               LastWeekTonight           24  2017-11-13T07:30:00.000Z   
2                  Rudy Mancuso           23  2017-11-12T19:05:24.000Z   
3         Good Mythical Morning           24  2017-11-13T11:00:04.000Z   
4                      nigahiga           24  2017-11-12T18:01:41.000Z   
...                         ...          ...                       ...   
375937      EnriqueIglesiasVEVO           10  2018-05-09T07:00:01.000Z   
375938          Jacob Sartorius           10  2018-05-11T17:09:16.000Z   
375939               Anne-Marie           10  2018-05-08T11:05:08.000Z   
375940  Eurovision Song Contest           24  2018-05-08T20:32:32.000Z   
375941           SuperDuperKyle           10  2018-05-11T04:06:35.000Z   

                                                     tags     views   likes  \
0                                         SHANtell martin    748374   57527   
1       last week tonight trump presidency|"last week ...   2418783   97185   
2       racist superman|"rudy"|"mancuso"|"king"|"bach"...   3191434  146033   
3       rhett and link|"gmm"|"good mythical morning"|"...    343168   10172   
4       ryan|"higa"|"higatv"|"nigahiga"|"i dare you"|"...   2095731  132235   
...                                                   ...       ...     ...   
375937  Enrique Iglesias feat. Pitbull|"MOVE TO MIAMI"...  25066952  268088   
375938  jacob sartorius|"jacob"|"up with it"|"jacob sa...   1492219   61998   
375939  anne|"marie"|"anne-marie"|"2002"|"two thousand...  29641412  394830   
375940  Eurovision Song Contest|"2018"|"Lisbon"|"Cypru...  14317515  151870   
375941  Kyle|"SuperDuperKyle"|"Ikuyo"|"2 Chainz"|"Soph...    607552   18271   

        dislikes  comment_count  \
0           2966          15954   
1           6146          12703   
2           5339           8181   
3            666           2146   
4           1989          17518   
...          ...            ...   
375937     12783           9933   
375938     13781          24330   
375939      8892          19988   
375940     45875          26766   
375941       274           1423   

                                        thumbnail_link  comments_disabled  \
0       https://i.ytimg.com/vi/2kyS6SvSYSE/default.jpg              False   
1       https://i.ytimg.com/vi/1ZAPwfrtAFY/default.jpg              False   
2       https://i.ytimg.com/vi/5qpjK5DgCt4/default.jpg              False   
3       https://i.ytimg.com/vi/puqaWrEC7tY/default.jpg              False   
4       https://i.ytimg.com/vi/d380meD0W0M/default.jpg              False   
...                                                ...                ...   
375937  https://i.ytimg.com/vi/l884

<pre style="color: black;">
files = glob.glob(os.path.join(path, "*videos.csv.zst"))
</pre>


This line of code allow us to finds all files ending with videos.csv.zst
- Uses glob.glob to search in the path given the structure that we are searching
- It returns a list of all files ending with videos.csv.zst.
- For example: ["ITvideos.csv.zst", "FRvideos.csv.zst", "DEvideos.csv.zst"].
---
<pre style="color: black;">
full_df = pd.concat((pd.read_csv(f,compression="zstd",encoding="utf-8",encoding_errors = "replace")
                       .assign(country=os.path.basename(f)[:2]) for f in files),
                       ignore_index=True)
</pre>

All files are read and concatenated into a single dataframe.
Going into specifics for each file f in the file list this portion of code is executed:

<pre style="color: black;">
pd.read_csv(f,compression="zstd",encoding="utf-8",encoding_errors = "replace")
</pre>

- compression="zstd" →  declaring that the file is compressed with zstandard (.zst)
- encoding="utf-8" → forces reading with this encoding to avoid errors
- encoding_errors = "replace" → when it encounters a line that it cannot encode, it replaces the character with �   (replacement character)


<pre style="color: black;">
.assign(country=os.path.basename(f)[:2])
</pre>

With this instructions a new column is created to insert a column called "country"
- os.path.basename(f)[:2] → extracting the name of the file (e.g. "ITvideos.csv.zst") selecting the first two characters
- .assign(country=...) → appending a new column called "country", and inserting the two character extracted before

In the final dataframe all the rows will have an added column with the country code
 
<pre style="color: black;">
pd.concat(..., ignore_index=True)
</pre>

This instruction permits to concatenate all the dataframe exctracted before
- Joins all the dataframe created by the cicle instructions
- ignore_index=True → resets all indexes
The result of this part of code is a unique dataframe with all the informations read in the files and with an aded column to specify the country.

---


<pre style="color: black;">
full_df
</pre>

Shows the dataframe, that has a shape of 375942 rows × 17 columns

---

## Exercise 2
### Extract all videos that have no tag.


In [4]:
no_tags = full_df[(full_df["tags"] == "[none]") |
                  (full_df["tags"].isna()) |
                  (full_df["tags"].str.strip() == "")]

<pre style="color: black;">
full_df["tags"] == "[none]"
</pre>

This portion of code ecxtracts all the videos that have in the column 'tags' a string "[none]"

--- 


<pre style="color: black;">
full_df["tags"].isna()
</pre>

Identify the row where the column 'tags' has a NaN value

---


<pre style="color: black;">
full_df["tags"].str.strip()
</pre>
All the rows with an empty tag column are selected: the instruction '.str' permit the use of string analysis and the strip() removes the spaces at the biginning  and at the endo of the string

Tho logic OR was used for this evaluation 

---

## Exercise 3
### For each channel, determine the total number of views


In [5]:
views_per_channel = (full_df.groupby("channel_title")["views"]
                            .sum()
                            .reset_index())
views_per_channel

channel_title    views
0                  ! 세상에 무슨일이  3942977
1               !!8時だよ面白ネタ大集合    50207
2              !BTS・TWICE まとめ     7310
3      !Los amorosos Virales¡     6069
4                     !t Live   240038
...                       ...      ...
37819              ５分でできる DIY   272607
37820                  ＢＡＳＨｔｖ   116292
37821           Ｋのフィッシングちゃんねる    37165
37822                 ＷＡＪＯＩＫＡ   424129
37823                💜Sandrea   347155

[37824 rows x 2 columns]

<pre style="color: black;">
full_df.groupby("channel_title")["views"]
</pre>
Grouping the dataframe for the name of the channel and selecting the views column

---
<pre style="color: black;">
.sum()
</pre>
For each channel compute the sum of the views

---
<pre style="color: black;">
.reset_index()
</pre>
Converts the index created by a groupby operation into regular columns, resulting in a standard tabular DataFrame.



---
<pre style="color: black;">
.sort_values(by="views", ascending=False)
</pre>
Sort channels from most viewed to least viewed

---

## Exercise 4
### Save all rows with disabled comments and disabled ratings, or that have video_error_or_removed in a new dataframe called excluded, and remove those rows from the original dataframe.


In [6]:
full_df[["comments_disabled","ratings_disabled","video_error_or_removed"]].dtypes

comments_disabled         bool
ratings_disabled          bool
video_error_or_removed    bool
dtype: object

<pre style="color: black;">
full_df[["comments_disabled","ratings_disabled","video_error_or_removed"]].dtypes
</pre>

Visualizing the type of the informations in the interested columns to understand witch kind of instruction use 

In [7]:
excluded = full_df[((full_df["comments_disabled"] == True) & (full_df["ratings_disabled"] == True)) |
                    (full_df["video_error_or_removed"] == True)].copy()
excluded

video_id trending_date  \
1435    Kn5UgGQukYQ      17.21.11   
1667    Kn5UgGQukYQ      17.22.11   
1914    Kn5UgGQukYQ      17.23.11   
2119    Kn5UgGQukYQ      17.24.11   
2203    RK_B4Ez4_5Q      17.25.11   
...             ...           ...   
375240  QW8whgmyTNU      18.10.06   
375408  QW8whgmyTNU      18.11.06   
375581  QW8whgmyTNU      18.12.06   
375756  QW8whgmyTNU      18.13.06   
375926  QW8whgmyTNU      18.14.06   

                                                    title       channel_title  \
1435    Breaking Bad's Bryan Cranston on Meeting Charl...  hudsonunionsociety   
1667    Breaking Bad's Bryan Cranston on Meeting Charl...  hudsonunionsociety   
1914    Breaking Bad's Bryan Cranston on Meeting Charl...  hudsonunionsociety   
2119    Breaking Bad's Bryan Cranston on Meeting Charl...  hudsonunionsociety   
2203    Verizon 360 Live: The Macy’s Thanksgiving Day ...             Verizon   
...                                                   ...                 ...   
375240                      This Is America: Women's Edit       Nicole Arbour   
375408                      This Is America: Women's Edit       Nicole Arbour   
375581                      This Is America: Women's Edit       Nicole Arbour   
375756                      This Is America: Women's Edit       Nicole Arbour   
375926                      This Is America: Women's Edit       Nicole Arbour   

        category_id              publish_time  \
1435              1  2016-10-14T21:14:51.000Z   
1667              1  2016-10-14T21:14:51.000Z   
1914              1  2016-10-14T21:14:51.000Z   
2119              1  2016-10-14T21:14:51.000Z   
2203             24  2017-11-23T19:51:07.000Z   
...             ...                       ...   
375240           24  2018-05-12T22:42:24.000Z   
375408           24  2018-05-12T22:42:24.000Z   
375581           24  2018-05-12T22:42:24.000Z   
375756           24  2018-05-12T22:42:24.000Z   
375926           24  2018-05-12T22:42:24.000Z   

                                                     tags    views  likes  \
1435    Breaking Bad|"Bryan Cranston"|"malcom in the m...    15058      0   
1667    Breaking Bad|"Bryan Cranston"|"malcom in the m...    34207      0   
1914    Breaking Bad|"Bryan Cranston"|"malcom in the m...    36137      0   
2119    Breaking Bad|"Bryan Cranston"|"malcom in the m...    36579      0   
2203    live stream|"360 video"|"fun videos for kids"|...  2618344  45197   
...                                                   ...      ...    ...   
375240  Nicole Arbour|"World's Sexiest Comedian"|"funn...  3125979      0   
375408  Nicole Arbour|"World's Sexiest Comedian"|"funn...  3130833      0   
375581  Nicole Arbour|"World's Sexiest Comedian"|"funn...  3140732      0   
375756  Nicole Arbour|"World's Sexiest Comedian"|"funn...  3149925      0   
375926  Nicole Arbour|"World's Sexiest Comedian"|"funn...  3158685      0   

        dislikes  comment_count  \
1435           0              0   
1667           0              0   
1914           0              0   
2119           0              0   
2203        2315           3332   
...          ...            ...   
375240         0              0   
375408         0              0   
375581         0              0   
375756         0              0   
375926         0              0   

                                        thumbnail_link  comments_disabled  \
1435    https://i.ytimg.com/vi/Kn5UgGQukYQ/default.jpg               True   
1667    https://i.ytimg.com/vi/Kn5UgGQukYQ/default.jpg               True   
1914    https://i.ytimg.com/vi/Kn5UgGQukYQ/default.jpg               True   
2119    https://i.ytimg.com/vi/Kn5UgGQukYQ/default.jpg               True   
2203    https://i.ytimg.com/vi/RK_B4Ez4_5Q/default.jpg              False   
...                                                ...                ...   
375240  https://i.ytimg.com/vi/QW8whgmyTNU/default.jpg               True   
375408  https://i.ytimg.com/vi/QW8whgmyTNU

<pre style="color: black;">
excluded = full_df[((full_df["comments_disabled"] == True) & (full_df["ratings_disabled"] == True)) |
                    (full_df["video_error_or_removed"] == True)].copy()
</pre>
This code selects all the videos that are going to be excluded form the analysis.
A video excluded has both comments and rating disabled or if it is a video classified as error or removed.
Conditions are combined using logic operators as AND or OR.
The result is a DataFrame of only the excluded videos.
The final Dataframe has 2620 rows and 17 columns.

---

In [8]:
full_df = full_df.drop(excluded.index)
full_df

video_id trending_date  \
0       2kyS6SvSYSE      17.14.11   
1       1ZAPwfrtAFY      17.14.11   
2       5qpjK5DgCt4      17.14.11   
3       puqaWrEC7tY      17.14.11   
4       d380meD0W0M      17.14.11   
...             ...           ...   
375937  l884wKofd54      18.14.06   
375938  IP8k2xkhOdI      18.14.06   
375939  Il-an3K9pjg      18.14.06   
375940  -DRsfNObKIQ      18.14.06   
375941  4YFo4bdMO8Q      18.14.06   

                                                    title  \
0                      WE WANT TO TALK ABOUT OUR MARRIAGE   
1       The Trump Presidency: Last Week Tonight with J...   
2       Racist Superman | Rudy Mancuso, King Bach & Le...   
3                        Nickelback Lyrics: Real or Fake?   
4                                I Dare You: GOING BALD!?   
...                                                   ...   
375937  Enrique Iglesias - MOVE TO MIAMI (Official Vid...   
375938  Jacob Sartorius - Up With It (Official Music V...   
375939                 Anne-Marie - 2002 [Official Video]   
375940  Eleni Foureira - Fuego - Cyprus - LIVE - First...   
375941  KYLE - Ikuyo feat.  2 Chainz & Sophia Black [A...   

                  channel_title  category_id              publish_time  \
0                  CaseyNeistat           22  2017-11-13T17:13:01.000Z   
1               LastWeekTonight           24  2017-11-13T07:30:00.000Z   
2                  Rudy Mancuso           23  2017-11-12T19:05:24.000Z   
3         Good Mythical Morning           24  2017-11-13T11:00:04.000Z   
4                      nigahiga           24  2017-11-12T18:01:41.000Z   
...                         ...          ...                       ...   
375937      EnriqueIglesiasVEVO           10  2018-05-09T07:00:01.000Z   
375938          Jacob Sartorius           10  2018-05-11T17:09:16.000Z   
375939               Anne-Marie           10  2018-05-08T11:05:08.000Z   
375940  Eurovision Song Contest           24  2018-05-08T20:32:32.000Z   
375941           SuperDuperKyle           10  2018-05-11T04:06:35.000Z   

                                                     tags     views   likes  \
0                                         SHANtell martin    748374   57527   
1       last week tonight trump presidency|"last week ...   2418783   97185   
2       racist superman|"rudy"|"mancuso"|"king"|"bach"...   3191434  146033   
3       rhett and link|"gmm"|"good mythical morning"|"...    343168   10172   
4       ryan|"higa"|"higatv"|"nigahiga"|"i dare you"|"...   2095731  132235   
...                                                   ...       ...     ...   
375937  Enrique Iglesias feat. Pitbull|"MOVE TO MIAMI"...  25066952  268088   
375938  jacob sartorius|"jacob"|"up with it"|"jacob sa...   1492219   61998   
375939  anne|"marie"|"anne-marie"|"2002"|"two thousand...  29641412  394830   
375940  Eurovision Song Contest|"2018"|"Lisbon"|"Cypru...  14317515  151870   
375941  Kyle|"SuperDuperKyle"|"Ikuyo"|"2 Chainz"|"Soph...    607552   18271   

        dislikes  comment_count  \
0           2966          15954   
1           6146          12703   
2           5339           8181   
3            666           2146   
4           1989          17518   
...          ...            ...   
375937     12783           9933   
375938     13781          24330   
375939      8892          19988   
375940     45875          26766   
375941       274           1423   

                                        thumbnail_link  comments_disabled  \
0       https://i.ytimg.com/vi/2kyS6SvSYSE/default.jpg              False   
1       https://i.ytimg.com/vi/1ZAPwfrtAFY/default.jpg              False   
2       https://i.ytimg.com/vi/5qpjK5DgCt4/default.jpg              False   
3       https://i.ytimg.com/vi/puqaWrEC7tY/default.jpg              False   
4       https://i.ytimg.com/vi/d380meD0W0M/default.jpg              False   
...                                                ...                ...   
375937  https://i.ytimg.com/vi/l884

<pre style="color: black;">
full_df = full_df.drop(excluded.index)
</pre>

Drops all rows from full_df using the indices contained in the excluded DataFrame.
The final DataFrame has 373322 rows and 17 columns

---

## Exercise 5
### Add a like_ratio column storing the ratio between the number of likes and of dislikes

In [9]:
full_df["like_ratio"] = full_df["likes"] / full_df["dislikes"].replace(0, pd.NA)
full_df

video_id trending_date  \
0       2kyS6SvSYSE      17.14.11   
1       1ZAPwfrtAFY      17.14.11   
2       5qpjK5DgCt4      17.14.11   
3       puqaWrEC7tY      17.14.11   
4       d380meD0W0M      17.14.11   
...             ...           ...   
375937  l884wKofd54      18.14.06   
375938  IP8k2xkhOdI      18.14.06   
375939  Il-an3K9pjg      18.14.06   
375940  -DRsfNObKIQ      18.14.06   
375941  4YFo4bdMO8Q      18.14.06   

                                                    title  \
0                      WE WANT TO TALK ABOUT OUR MARRIAGE   
1       The Trump Presidency: Last Week Tonight with J...   
2       Racist Superman | Rudy Mancuso, King Bach & Le...   
3                        Nickelback Lyrics: Real or Fake?   
4                                I Dare You: GOING BALD!?   
...                                                   ...   
375937  Enrique Iglesias - MOVE TO MIAMI (Official Vid...   
375938  Jacob Sartorius - Up With It (Official Music V...   
375939                 Anne-Marie - 2002 [Official Video]   
375940  Eleni Foureira - Fuego - Cyprus - LIVE - First...   
375941  KYLE - Ikuyo feat.  2 Chainz & Sophia Black [A...   

                  channel_title  category_id              publish_time  \
0                  CaseyNeistat           22  2017-11-13T17:13:01.000Z   
1               LastWeekTonight           24  2017-11-13T07:30:00.000Z   
2                  Rudy Mancuso           23  2017-11-12T19:05:24.000Z   
3         Good Mythical Morning           24  2017-11-13T11:00:04.000Z   
4                      nigahiga           24  2017-11-12T18:01:41.000Z   
...                         ...          ...                       ...   
375937      EnriqueIglesiasVEVO           10  2018-05-09T07:00:01.000Z   
375938          Jacob Sartorius           10  2018-05-11T17:09:16.000Z   
375939               Anne-Marie           10  2018-05-08T11:05:08.000Z   
375940  Eurovision Song Contest           24  2018-05-08T20:32:32.000Z   
375941           SuperDuperKyle           10  2018-05-11T04:06:35.000Z   

                                                     tags     views   likes  \
0                                         SHANtell martin    748374   57527   
1       last week tonight trump presidency|"last week ...   2418783   97185   
2       racist superman|"rudy"|"mancuso"|"king"|"bach"...   3191434  146033   
3       rhett and link|"gmm"|"good mythical morning"|"...    343168   10172   
4       ryan|"higa"|"higatv"|"nigahiga"|"i dare you"|"...   2095731  132235   
...                                                   ...       ...     ...   
375937  Enrique Iglesias feat. Pitbull|"MOVE TO MIAMI"...  25066952  268088   
375938  jacob sartorius|"jacob"|"up with it"|"jacob sa...   1492219   61998   
375939  anne|"marie"|"anne-marie"|"2002"|"two thousand...  29641412  394830   
375940  Eurovision Song Contest|"2018"|"Lisbon"|"Cypru...  14317515  151870   
375941  Kyle|"SuperDuperKyle"|"Ikuyo"|"2 Chainz"|"Soph...    607552   18271   

        dislikes  comment_count  \
0           2966          15954   
1           6146          12703   
2           5339           8181   
3            666           2146   
4           1989          17518   
...          ...            ...   
375937     12783           9933   
375938     13781          24330   
375939      8892          19988   
375940     45875          26766   
375941       274           1423   

                                        thumbnail_link  comments_disabled  \
0       https://i.ytimg.com/vi/2kyS6SvSYSE/default.jpg              False   
1       https://i.ytimg.com/vi/1ZAPwfrtAFY/default.jpg              False   
2       https://i.ytimg.com/vi/5qpjK5DgCt4/default.jpg              False   
3       https://i.ytimg.com/vi/puqaWrEC7tY/default.jpg              False   
4       https://i.ytimg.com/vi/d380meD0W0M/default.jpg              False   
...                                                ...                ...   
375937  https://i.ytimg.com/vi/l884

<pre style="color: black;">
full_df["like_ratio"] = full_df["likes"] / full_df["dislikes"].replace(0, pd.NA)
</pre>
Create a new columns computing the ratio between likes and dislikes

<pre style="color: black;">
full_df["dislikes"].replace(0, pd.NA)
</pre>
This instruction controll the case where 'dislike' is zero, in this case it replaces it with a NaN value to solve eventual problems with the division.

---

## Exercise 6
### Cluster the publish time into 10-minute intervals (e.g. from 02:20 to 02:30)

In [10]:
full_df["publish_time"] = pd.to_datetime(full_df["publish_time"], errors="coerce")
g = (full_df["publish_time"].dt.minute // 10) * 10
start  = full_df["publish_time"].dt.floor("10min")
end = start + pd.Timedelta(minutes = 10)
full_df["publish_interval"] = ("from " + start.dt.strftime("%H:%M") + " to " + end.dt.strftime("%H:%M"))
full_df

video_id trending_date  \
0       2kyS6SvSYSE      17.14.11   
1       1ZAPwfrtAFY      17.14.11   
2       5qpjK5DgCt4      17.14.11   
3       puqaWrEC7tY      17.14.11   
4       d380meD0W0M      17.14.11   
...             ...           ...   
375937  l884wKofd54      18.14.06   
375938  IP8k2xkhOdI      18.14.06   
375939  Il-an3K9pjg      18.14.06   
375940  -DRsfNObKIQ      18.14.06   
375941  4YFo4bdMO8Q      18.14.06   

                                                    title  \
0                      WE WANT TO TALK ABOUT OUR MARRIAGE   
1       The Trump Presidency: Last Week Tonight with J...   
2       Racist Superman | Rudy Mancuso, King Bach & Le...   
3                        Nickelback Lyrics: Real or Fake?   
4                                I Dare You: GOING BALD!?   
...                                                   ...   
375937  Enrique Iglesias - MOVE TO MIAMI (Official Vid...   
375938  Jacob Sartorius - Up With It (Official Music V...   
375939                 Anne-Marie - 2002 [Official Video]   
375940  Eleni Foureira - Fuego - Cyprus - LIVE - First...   
375941  KYLE - Ikuyo feat.  2 Chainz & Sophia Black [A...   

                  channel_title  category_id              publish_time  \
0                  CaseyNeistat           22 2017-11-13 17:13:01+00:00   
1               LastWeekTonight           24 2017-11-13 07:30:00+00:00   
2                  Rudy Mancuso           23 2017-11-12 19:05:24+00:00   
3         Good Mythical Morning           24 2017-11-13 11:00:04+00:00   
4                      nigahiga           24 2017-11-12 18:01:41+00:00   
...                         ...          ...                       ...   
375937      EnriqueIglesiasVEVO           10 2018-05-09 07:00:01+00:00   
375938          Jacob Sartorius           10 2018-05-11 17:09:16+00:00   
375939               Anne-Marie           10 2018-05-08 11:05:08+00:00   
375940  Eurovision Song Contest           24 2018-05-08 20:32:32+00:00   
375941           SuperDuperKyle           10 2018-05-11 04:06:35+00:00   

                                                     tags     views   likes  \
0                                         SHANtell martin    748374   57527   
1       last week tonight trump presidency|"last week ...   2418783   97185   
2       racist superman|"rudy"|"mancuso"|"king"|"bach"...   3191434  146033   
3       rhett and link|"gmm"|"good mythical morning"|"...    343168   10172   
4       ryan|"higa"|"higatv"|"nigahiga"|"i dare you"|"...   2095731  132235   
...                                                   ...       ...     ...   
375937  Enrique Iglesias feat. Pitbull|"MOVE TO MIAMI"...  25066952  268088   
375938  jacob sartorius|"jacob"|"up with it"|"jacob sa...   1492219   61998   
375939  anne|"marie"|"anne-marie"|"2002"|"two thousand...  29641412  394830   
375940  Eurovision Song Contest|"2018"|"Lisbon"|"Cypru...  14317515  151870   
375941  Kyle|"SuperDuperKyle"|"Ikuyo"|"2 Chainz"|"Soph...    607552   18271   

        dislikes  comment_count  \
0           2966          15954   
1           6146          12703   
2           5339           8181   
3            666           2146   
4           1989          17518   
...          ...            ...   
375937     12783           9933   
375938     13781          24330   
375939      8892          19988   
375940     45875          26766   
375941       274           1423   

                                        thumbnail_link  comments_disabled  \
0       https://i.ytimg.com/vi/2kyS6SvSYSE/default.jpg              False   
1       https://i.ytimg.com/vi/1ZAPwfrtAFY/default.jpg              False   
2       https://i.ytimg.com/vi/5qpjK5DgCt4/default.jpg              False   
3       https://i.ytimg.com/vi/puqaWrEC7tY/default.jpg              False   
4       https://i.ytimg.com/vi/d380meD0W0M/default.jpg              False   
...                                                ...                ...   
375937  https://i.ytimg.com/vi/l884

<pre style="color: black;">
full_df["publish_time"] = pd.to_datetime(full_df["publish_time"], errors="coerce")
</pre>

This instruction converts 'publish_time' in a date-time format, fixing each timezone, the error correction sets all the invalid data in NaT

---
<pre style="color: black;">
g = (full_df["publish_time"].dt.minute // 10) * 10
</pre>
full_df["publish_time"].dt.minute extracts only the minutes of the time (e.g. 14:23:10 returns only 23)

Having the minutage, it divide it by 10, and extracts only the integer part, so we are able to understand in witch cluster belongs. 
The code multimplies for a 10 factor so it extracts the cluster biginning minutage

---
<pre style="color: black;">
start  = full_df["publish_time"].dt.floor("10min")
</pre>
This variable collects the starting time of the cluster 

<pre style="color: balck;">
end = start + pd.Timedelta(minutes = 10)
</pre>

An object with the duration of the cluster is created (10 min) and it is added to the start variable, so we have also a variable with the end time of the cluster

---
We want the fesult to be form xx:xx to xx:xx, so the following instructions were executed

<pre style="color: black;">
full_df["publish_interval"] = ("from " + start.dt.strftime("%H:%M") + " to "
                               + end.dt.strftime("%H:%M"))
</pre>

We transform a data datetime into a string with a format HH:MM, and then we add them to the strings 'form' and 'to'

---

# ESERCIZIO 7
### For each interval, determine the number of videos, average number of likes and of dislikes.

In [11]:
videos_count = full_df.groupby("publish_interval")["video_id"].count()
avg_likes = full_df.groupby("publish_interval")["likes"].mean()
avg_dislikes = full_df.groupby("publish_interval")["dislikes"].mean()
df_intervals_stats = pd.DataFrame({"videos count" : videos_count,
                                   "avg_likes" : avg_likes,
                                   "avg_dislikes" : avg_dislikes})           
df_intervals_stats

videos count     avg_likes  avg_dislikes
publish_interval                                             
from 00:00 to 00:10          2897  61288.115637   3808.149465
from 00:10 to 00:20          1509  22748.138502   1449.836315
from 00:20 to 00:30          1241  21378.280419   1072.344883
from 00:30 to 00:40          1614  36853.560719    955.890954
from 00:40 to 00:50          1269  42198.623325   1909.301812
...                           ...           ...           ...
from 23:10 to 23:20          1554  22783.840412   1110.168597
from 23:20 to 23:30          1438  30696.510431   1177.020862
from 23:30 to 23:40          1666  18159.173469    874.184874
from 23:40 to 23:50          1466  45736.581855   9341.549795
from 23:50 to 00:00          1269  10853.743893    505.701340

[144 rows x 3 columns]

<pre style="color: black;">
videos_count = full_df.groupby("publish_interval")["video_id"].count()
</pre>
Create a Series grouping by the column 'publish_interval' and counts the total videos for each cluster

---
<pre style="color: black;">
avg_likes = full_df.groupby("publish_interval")["likes"].mean()
</pre>

Create a Series grouping by the column 'publish_interval' and compute the likes' mean for each cluster

---
<pre style="color: black;">
avg_likes = full_df.groupby("publish_interval")["dislikes"].mean()
</pre>

Create a Series grouping by the column 'publish_interval' and compute the dislikes' mean for each cluster

---
<pre style="color: black;">
df_intervals_stats = pd.DataFrame({"videos count" : videos_count,
                                   "avg_likes" : avg_likes,
                                   "avg_dislikes" : avg_dislikes})  
</pre>
Creates a new dataframe where all the informations collected  

---

# ESERCIZIO 8
### For each tag, determine the number of videos. Notice that tags contains a string with several tags.

---


<pre style="color: red;">
full_df["tags"] = full_df["tags"].replace("[none]", "")
</pre>


---


sostituisce la stringa esatta "[none]" con una stringa vuota

In [12]:
full_df["tags_list"] = full_df["tags"].str.split("|")
tags_exploded = full_df.explode("tags_list")
tags_exploded["tags_list"] = (tags_exploded["tags_list"].str.replace('"', '', regex=False)
                                                        .str.lower()
                                                        .replace("none","[none]"))
tag_counts = (tags_exploded.groupby("tags_list")["video_id"]
                           .count()
                           .reset_index(name="video_count")
                           .sort_values("video_count", ascending=False)).reset_index().drop(columns=["index"])
tag_counts = tag_counts[tag_counts["tags_list"] != "[none]"]
tag_counts

tags_list  video_count
1                              funny        17308
2                             comedy        15695
3                               2018        11371
4                               news         9132
5                              music         8196
...                              ...          ...
767983                        snowie            1
767984              cadbury mini egg            1
767985        cadbury mini eggs cake            1
767986                snowflake fail            1
767987  🤝 zarbex zahlt schutzgeld!!!            1

[767987 rows x 2 columns]

<pre style="color: black;">
full_df["tags_list"] = full_df["tags"].str.split("|")
</pre>

Dividing each string of the 'tags' column using the character "|" as separator

---
<pre style="color: black;">
tags_exploded = full_df.explode("tags_list")
</pre>
The instruction explode creates separates in single tags each tug in the column 'tags_list', creating new rows for each element 

---
<pre style="color: black;">
tags_exploded = tags_exploded[tags_exploded["tags_list"] != ""]
</pre>
Selecting only the rows where there are some tags, excluding the empty ones (e.g. "")

---
<pre style="color: black;">
tag_counts = (tags_exploded.groupby("tags_list")["video_id"]
                           .count()
                           .reset_index(name="video_count")
                           .sort_values("video_count", ascending=False)).reset_index().drop(columns=["index"])
</pre>
Gropuing all the lines with the same tag and counting how many times each tag is used, sorting the values in a descending order and ressetting indexes for the new Dataframe, removing the indexes created by the groupby

---
<pre style="color: black;">
tag_counts = tag_counts[tag_counts["tags_list"] != "[none]"]
</pre>
Excluding the tags that are "[none]"

---

## Exercise 9
### Find the tags with the largest number of videos

In [13]:
max_count = tag_counts['video_count'].max()
max_tag = tag_counts[tag_counts["video_count"] == max_count]
max_tag

tags_list  video_count
1     funny        17308

<pre style="color: black;">
max_count = tag_counts['video_count'].max()
</pre>
Searching for the max value in the tag_counts Dataframe

---
<pre style="color: black;">
max_tag = tag_counts[tag_counts["video_count"] == max_count]
</pre>
Searching for the tags that have the same count of the max_count varible

---

# Exercise 10
### For each (tag, country) pair, compute average ratio likes/dislikes

In [14]:
tag_country_ratio = (tags_exploded.groupby(["tags_list", "country"])["like_ratio"]
                                  .mean()
                                  .reset_index(name="avg_like_ratio")
                                  .sort_values("avg_like_ratio", ascending=False)
                                  .fillna("dislike = 0"))

tag_country_ratio.reset_index().drop(columns=["index"])

tags_list country avg_like_ratio
0                u-verse      RU        11688.0
1               directtv      RU        11688.0
2       audience network      RU        11688.0
3                   at&t      RU        11688.0
4          at&t audience      RU        11688.0
...                  ...     ...            ...
999173         ｂｇ～身辺警護人～      JP    dislike = 0
999174                ｂｌ      JP    dislike = 0
999175              ｆｃ東京      JP    dislike = 0
999176        ｇａｎｇｐａｒａｄｅ      JP    dislike = 0
999177    ｊリーグybcルヴァンカップ      JP    dislike = 0

[999178 rows x 3 columns]

<pre style="color: black;">
tag_country_ratio = (tags_exploded.groupby(["tags_list", "country"])["like_ratio"]
                                  .mean()
                                  .reset_index(name="avg_like_ratio")
                                  .sort_values("avg_like_ratio", ascending=False)
                                  .fillna("dislike = 0"))
</pre>
Grouping the Dataframe for tag and country and computing the mean of 'like_ratio', renaming the column as 'avg_like_ratio' and sorting the rows for descending values.
If dislike is equal to 0, and the fraction will give a NaN value, in the column will appear "dislike = 0"

---
<pre style="color: black;">
tag_country_ratio.reset_index().drop(columns=["index"])
</pre>
Dropping the original index column and resetting it

# Exercise 11
### For each (trending_date, country) pair, the video with the largest number of views

In [15]:
idx = full_df.groupby(["trending_date", "country"])["views"].idxmax()
top_videos = full_df.loc[idx]
top_videos

video_id trending_date  \
258804  6ZfuNTqbHE8      17.01.12   
217965  6ZfuNTqbHE8      17.01.12   
44480   3VbHg5fqBYw      17.01.12   
340613  TyHvyGVs42U      17.01.12   
125380  6ZfuNTqbHE8      17.01.12   
...             ...           ...   
212122  aJOTlE1K90k      18.31.05   
191438  ntWSnDV0MYs      18.31.05   
119353  xTlNMmZKwpA      18.31.05   
334091  1nX0kF2UwDc      18.31.05   
38146   VYOjWnS4cMY      18.31.05   

                                                    title  \
258804  Marvel Studios' Avengers: Infinity War Officia...   
217965  Marvel Studios' Avengers: Infinity War Officia...   
44480                Avengers: Infinity War Trailer Tease   
340613          Luis Fonsi, Demi Lovato - Échame La Culpa   
125380  Marvel Studios' Avengers: Infinity War Officia...   
...                                                   ...   
212122              Maroon 5 - Girls Like You ft. Cardi B   
191438                     [MV] AOA _ Bingle Bangle(빙글뱅글)   
119353  Cardi B, Bad Bunny & J Balvin - I Like It [Off...   
334091  [BadComedian] - Движение Вверх (Плагиат или ве...   
38146   Childish Gambino - This Is America (Official V...   

               channel_title  category_id              publish_time  \
258804  Marvel Entertainment           24 2017-11-29 13:26:24+00:00   
217965  Marvel Entertainment           24 2017-11-29 13:26:24+00:00   
44480   Marvel Entertainment           24 2017-11-28 17:09:22+00:00   
340613         LuisFonsiVEVO           10 2017-11-17 05:00:01+00:00   
125380  Marvel Entertainment           24 2017-11-29 13:26:24+00:00   
...                      ...          ...                       ...   
212122           Maroon5VEVO           10 2018-05-31 04:00:00+00:00   
191438          1theK (원더케이)           10 2018-05-28 09:00:03+00:00   
119353               Cardi B           10 2018-05-29 14:05:10+00:00   
334091           BadComedian           23 2018-05-30 18:20:48+00:00   
38146    ChildishGambinoVEVO           10 2018-05-06 04:00:07+00:00   

                                                     tags      views    likes  \
258804  marvel|"comics"|"comic books"|"nerdy"|"geeky"|...   56367282  2157741   
217965  marvel|"comics"|"comic books"|"nerdy"|"geeky"|...   56367282  2157737   
44480   marvel"|"comics"|"comic books"|"nerdy"|"geeky"...    7281189   180808   
340613  Luis|"Fonsi"|"Demi"|"Lovato"|"Échame"|"La"|"Cu...  143408235  2686169   
125380  marvel|"comics"|"comic books"|"nerdy"|"geeky"|...   56367282  2157733   
...                                                   ...        ...      ...   
212122  Maroon|"Girls"|"Like"|"You"|"Interscope"|"Reco...    3057987   406732   
191438  Kpop|"1theK"|"원더케이"|"loen"|"로엔"|"뮤비"|"티져"|"MV"...    4150448   186262   
119353  Cardi B|"I Like It"|"Invasion of Privacy"|"Bad...   20723565  1018817   
334091  BadComedian|"Евген"|"Bad"|"Comedian"|"Фильмы"|...    3125598   334685   
38146   Childish Gambino|"Rap"|"This Is America"|"mcDJ...  217750076  4934188   

        dislikes  comment_count  \
258804     34078         303178   
217965     34077         303178   
44480      19955          21244   
340613    137938         144217   
125380     34077         303178   
...          ...            ...   
212122      4521          31301   
191438      9728          15989   
119353     48091          68790   
334091      7777          34556   
38146     335462         509799   

                                        thumbnail_link  comments_disabled  \
258804  https://i.ytimg.com/vi/6ZfuNTqbHE8/default.jpg              False   
217965  https://i.ytimg.com/vi/6ZfuNTqbHE8/default.jpg              False   
44480   https://i.ytimg.com/vi/3VbHg5fqBYw/default.jpg              False   
340613  https://i.ytimg.com/vi/TyHvyGVs42U/default.jpg              False   
125380  https://i.ytimg.com/vi/6ZfuNTqbHE8/default.jpg              False   
...                                                ...                ...   
212122  https://i.ytimg.com/vi/aJOTlE1K90k/defa

<pre style="color: black;">
idx = full_df.groupby(["trending_date", "country"])["views"].idxmax()
</pre>
Grouping the videos by trending date and country and searching for the index of the one with the most views

---
<pre style="color: black;">
top_videos = full_df.loc[idx]
</pre>
Selecting all the videos with the same indexes collected in the instruction before

---

# ESERCIZIO 12
### Divide trending_date into three columns: year, month, day

In [16]:
full_df["trending_date"] = pd.to_datetime(full_df["trending_date"],format="%y.%d.%m",errors="coerce")
full_df["year"] = full_df["trending_date"].dt.year
full_df["month"] = full_df["trending_date"].dt.month
full_df["day"] = full_df["trending_date"].dt.day

<pre style="color: black;">
full_df["trending_date"] = pd.to_datetime(full_df["trending_date"],format="%y.%d.%m",errors="coerce")
</pre>
Converts the 'trending_date' column to a pandas datetime type using an explicit two-digit year / day / month format (YY.DD.MM), coercing invalid or non-matching values to NaT to ensure consistent and unambiguous date parsing.

---
<pre style="color: black;">
full_df["year"] = full_df["trending_date"].dt.year
full_df["month"] = full_df["trending_date"].dt.month
full_df["day"] = full_df["trending_date"].dt.day
</pre>
Creating three new columns with year, month and day exctracted from the 'trending_date' column

---

# Exercise 13
### For each (month, country) pair, the video with the largest number of views

In [17]:
idx_df = full_df.groupby(["month", "country"])["views"].idxmax()
top_month_country = full_df.loc[idx_df]
top_month_country.shape

(77, 23)

<pre style="color: black;">
idx_df = full_df.groupby(["month", "country"])["views"].idxmax()
</pre>
Grouping for each month and country and selecting the indexes of the videos with the most view

---
<pre style="color: black;">
top_month_country = full_df.loc[idx_df].sort_values(["views"])
</pre>

Selecting all the rows of with the indexes found and sorting them by the number of views

---

# ESERCIZIO 14
### Read all json files with the video categories

In [18]:
json_files = glob.glob(os.path.join(path, "*_category_id.json"))
categories_list = []

for f in json_files:
    country = os.path.basename(f)[:2].upper()
    
    with open(f, "r", encoding="utf-8") as file:
        data = json.load(file)

    for item in data["items"]:
        categories_list.append({"category_id": int(item["id"]),
                                "category_title": item["snippet"]["title"],
                                "country": country,
                                "assignable": item["snippet"].get("assignable", False)})

categories_df = pd.DataFrame(categories_list)
categories_df

category_id    category_title country  assignable
0              1  Film & Animation      IN        True
1              2  Autos & Vehicles      IN        True
2             10             Music      IN        True
3             15    Pets & Animals      IN        True
4             17            Sports      IN        True
..           ...               ...     ...         ...
306           40    Sci-Fi/Fantasy      DE       False
307           41          Thriller      DE       False
308           42            Shorts      DE       False
309           43             Shows      DE       False
310           44          Trailers      DE       False

[311 rows x 4 columns]

<pre style="color: black;">
json_files = glob.glob(os.path.join(path, "*_category_id.json"))
</pre>


This line of code allow us to finds all files ending with _category_id.json
- Uses glob.glob to search in the path given the structure that we are searching
- Uses os library to choose the pattern where to search, joining th path and the structure of interest
- It returns a list of all files ending with _category_id.json
---
<pre style="color: black;">
for f in json_files:
    country = os.path.basename(f)[:2].upper()
    
    with open(f, "r", encoding="utf-8") as file:
        data = json.load(file)

    for item in data["items"]:
        categories_list.append({"category_id": int(item["id"]),
                                "category_title": item["snippet"]["title"],
                                "country": country,
                                "assignable": item["snippet"].get("assignable", False)})
</pre>

For each file the code exctracts different informations
Going into specifics for each file f in the file list this portion of code is executed:

<pre style="color: black;">
country = os.path.basename(f)[:2].upper()
</pre>

- os.path.basename(f) takes only the filename (removing the directory path),
- [:2] selects the first two characters that represent the country code,
- .upper() converts them to uppercase (e.g., "it" → "IT").


<pre style="color: black;">
with open(f, "r", encoding="utf-8") as file:
        data = json.load(file)
</pre>

Opens the JSON file in read mode using UTF-8 encoding and assigns it to the variable file.
The with statement ensures the file is automatically closed after reading.
Then parses the JSON content of the file and loads it into a Python dictionary called data
 
<pre style="color: black;">
for item in data["items"]:
        categories_list.append({"category_id": int(item["id"]),
                                "category_title": item["snippet"]["title"],
                                "country": country,
                                "assignable": item["snippet"].get("assignable", False)})
</pre>

Iterates over all category entries in the JSON file and appends a structured dictionary to categories_list for each item, extracting the category ID, human-readable title, associated country code, and whether the category is assignable to videos (defaulting to False if the field is missing).

---


<pre style="color: black;">
categories_df = pd.DataFrame(categories_list)
categories_df
</pre>
Converts the list of category dictionaries into a pandas DataFrame, where each dictionary becomes a row and each key becomes a column.
Shows the dataframe, that has a shape of 311 rows × 4 columns

---


# ESERCIZIO 15
### For each country, determine how many videos have a category that is not assignable.

In [19]:
merged = full_df.merge(categories_df,how="left",
                       left_on=["category_id", "country"],
                       right_on=["category_id", "country"])
not_assigable = merged[merged['assignable'] == False]
not_assignable_per_country = (not_assigable.groupby("country")["video_id"]
                                            .nunique()
                                            .reset_index(name="not_assignable_videos"))
not_assignable_per_country

country  not_assignable_videos
0      CA                    111
1      DE                    105
2      FR                    107
3      GB                      1
4      IN                    124
5      KR                     75
6      MX                      3
7      RU                    169
8      US                      4

<pre style="color: black;">
merged = full_df.merge(categories_df,how="left",
                       left_on=["category_id", "country"],
                       right_on=["category_id", "country"])
</pre>
Performs a left join between the video-level dataset (full_df) and the category metadata table (categories_df), matching rows on both 'category_id' and 'country'. All rows from full_df are preserved, and category information
is added where a corresponding match exists

---
<pre style="color: black;">
not_assigable = merged[merged['assignable'] == False]
</pre>
Filters the merged dataset to retain only videos whose category is not assignable

---
<pre style="color: black;">
not_assignable_per_country = (not_assigable.groupby("country")["video_id"]
                                            .nunique()
                                            .reset_index(name="not_assignable_videos"))
</pre>
Groups the filtered videos by country and counts the number of unique videos (using 'video_id') with a non-assignable category in each country

---

In [20]:
not_assignable_per_country = (not_assigable.groupby("country")['assignable']
                                           .count()
                                           .reset_index()
                                           .rename(columns = {"assignable":"not_assignable"}))
not_assignable_per_country

country  not_assignable
0      CA             130
1      DE             110
2      FR             112
3      GB              20
4      IN             221
5      KR             167
6      MX               3
7      RU             195
8      US              57